keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,514 @@
+   * Retain - this is the default policy and is considered in cases where no policy is specified. 
+       This would be the existing behaviour - when a StatefulSet is deleted, no action is taken with
+       respect to the PVCs created by the StatefulSet.
+   * RemoveOnScaledown - When a pod is deleted on scale down, the corresponding PVC is deleted as well. 
+       A scale up following a scale down, will wait till old PVC for the removed Pod is deleted and ensure 
+       that the PVC used is a freshly created one.
+   * RemoveOnStatefulSetDeletion - PVCs corresponding to the StatefulSet are deleted when StatefulSet
+       themselves get deleted.
+3. Add `patch` to the statefulset controller rbac cluster role for `persistentvolumeclaims`.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+User environment is such at the content of the PVCs which are created automatically during StatefulSet 
+creation need not be retained after the StatefulSet is deleted. User also requires that the scale 
+up/down occurs in a fast manner, and leverages any previously existing auto created PVCs within the 
+life time of the StatefulSet. An option needs to be provided for the user to auto-delete the PVCs 
+once the StatefulSet is deleted. 
+
+User would set the VolumeReclaimPolicy as 'RemoveOnStatefulSetDelete' which would ensure that 
+the PVCs created automatically during the StatefulSet activation is removed once the StatefulSet 
+is deleted.
+
+#### Story 2
+User is cost conscious but at the same time can sustain slower scale up(after a scale down) speeds. Needs 
+a provision where the PVC created for a pod(which is part of the StatefulSet) is removed when the Pod 
+is deleted as part of a scale down. Since the subsequent scale up needs to create fresh PVCs, it will 
+be slower than scale ups relying on existing PVCs(from earlier scale ups). 
+
+User would set the VolumeReclaimPolicy as 'RemoveOnScaledown' ensuring PVCs are deleted when corresponding
+Pods are deleted. New Pods created during sclae up followed by a scaledown will wait for freshly created PVCs.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
A discussion of durability risks with the changed behavior (even though it is opt-in) should be added here.

Reply:
@kk-src Maybe something like the following?

If the PVCReclaimPolicy is changed from its default of "Retain", then PVs will be deleted on SS scaledown or deletion. A user will lose data on those PVs if proactive effort is not taken to replicate that data.  However, PVs associated with the StatefulSet will be more durable than ephemeral volumes would be, as they are only deleted on scaledown or StatefulSet deletion, and not on other pod lifecycle events like being rescheduled to a new node, even with the new retain policies.